htlcswitch/link.go

```diff
@@ -373,8 +397,30 @@ type channelLink struct {
 	downstream chan *htlcPacket
 
 	// shutdownRequest is a channel that the channelLink will listen on to
-	// service shutdown requests from ShutdownIfChannelClean calls.
-	shutdownRequest chan *shutdownReq
+	// service shutdown requests from NotifyShouldShutdown calls.
+	shutdownRequest chan *chanCloseCtx
+
+	// localCloseReq stores the local ChanClose request to pass to the
+	// peer.Brontide when we're done with the Shutdown phase.
+	localCloseReq *ChanClose
+
+	// shutdownInit is a bool that is set when we've initiated a coop
+	// close.
+	shutdownInit bool
+
+	// shutdownReceived is a bool that is set when we've received a
+	// Shutdown message from the remote peer.
+	//
+	// The mutex is only accessed if:
+	// - a write is occurring
+	// - a read is occurring from hasReceivedShutdown
+	// all reads from the htlcManager goroutine should be race-free.
```

Comment:
The mutex is only accessed in production code if a write occurs. This happens in `handleUpstreamMsg`. The `hasReceivedShutdown` function is only called in test code. If we make this an `atomic.Bool`, then all of the read-only code that previously didn't need to access the mutex (since the reads are in the same goroutine as the write except for the test-only `hasReceivedShutdown`) will need to now use atomics. Is there that much of a noticeable difference in performance? If so, will change